staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/keystone/keystone.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keystone
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/utils"
+
+	"fmt"
+	netutil "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apiserver/pkg/authentication/user"
+	certutil "k8s.io/client-go/util/cert"
+)
+
+// KeystoneAuthenticator contacts openstack keystone to validate user's token passed in the request.
+// The keystone endpoint is passed during apiserver startup
+type KeystoneAuthenticator struct {
+	authURL string
+	client  *gophercloud.ServiceClient
+}
+
+// AuthenticatePassword checks the token via Keystone call
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+
+	// We can use the Keystone GET /v3/auth/tokens API to validate the token
+	// and get information about the user as well
+	// http://git.openstack.org/cgit/openstack/keystone/tree/api-ref/source/v3/authenticate-v3.inc#n437
+	// https://developer.openstack.org/api-ref/identity/v3/?expanded=validate-and-show-information-for-token-detail
+	request_opts := gophercloud.RequestOpts{
+		MoreHeaders: map[string]string{
+			"X-Auth-Token":    token,
+			"X-Subject-Token": token,
+		},
+	}
+	url := keystoneAuthenticator.client.ServiceURL("auth", "tokens")
+	response, err := keystoneAuthenticator.client.Request("GET", url, &request_opts)
+	if err != nil {
+		glog.V(4).Info("Failed: bad response from API call: %v", err)
+		return nil, false, errors.New("Failed to authenticate")
+	}
+
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(4).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, errors.New("Failed to authenticate")
+	}
+
+	// Read the response and parse the name and id for the user
+	var data map[string]interface{}
+	err = json.Unmarshal(bodyBytes, &data)
+	if err != nil {
+		glog.V(4).Infof("Cannot unmarshal response: %v", err)
+		return nil, false, errors.New("Failed to authenticate")
+	}
+
+	token_info := data["token"].(map[string]interface{})
+	project_info := token_info["project"].(map[string]interface{})
+	user_info := &user.DefaultInfo{
+		Name: project_info["name"].(string),
```

Comment:
This raises the same questions as previous keystone PRs about how to map keystone user and project info to kubernetes user info. Given that there is not a single way to map this info, I don't think it is wise to bless this one mapping over others. I'd rather see keystone token authentication externalized on the server side. 

On the client side, a keystone auth provider for kubectl that obtains a token based on envvars seems like it would fit well with both the kube and openstack ecosystems